arb_os/arbretryable.mini

```diff
@@ -0,0 +1,234 @@
+use evmCallStack::evmCallFrame_getCalldata;
+use evmCallStack::evmCallFrame_getCaller;
+use evmCallStack::evmCallFrame_getCallvalue;
+use evmCallStack::evmCallFrame_getMemory;
+use evmCallStack::evmCallStack_callHitError;
+use evmCallStack::evmCallStack_getAccountInCurrentContext;
+use evmCallStack::evmCallStack_setAccountInCurrentContext;
+use evmCallStack::getRetryBufferInCurrentContext;
+use evmCallStack::setRetryBufferInCurrentContext;
+
+use evmOps::evmOp_log2;
+use evmOps::evmOp_return;
+use evmOps::evmOp_revert_knownPc;
+
+use std::bytearray::ByteArray;
+use std::bytearray::bytearray_new;
+use std::bytearray::bytearray_size;
+use std::bytearray::bytearray_get256;
+use std::bytearray::bytearray_set256;
+
+use retrybuffer::retryBuffer_getLifetime;
+use retrybuffer::retryBuffer_getTxTimeout;
+use retrybuffer::redeemRetryableTxInTx;
+use retrybuffer::retryBuffer_getTxKeepalivePrice;
+use retrybuffer::retryBuffer_keepalive;
+use retrybuffer::retryBuffer_getBeneficiary;
+use retrybuffer::retryBuffer_delete;
+
+
+public impure func arbretryable_txcall() {
+    if let Some(topFrame) = evmCallStack_topFrame() {
+        let calldata = evmCallFrame_getCalldata(topFrame);
+        if (bytearray_size(calldata) < 4) {
+            evmOp_revert_knownPc(0, 0, 0);
+        }
+        let funcCode = asm(224, bytearray_get256(calldata, 0)) uint { shr };
+        if (funcCode == const::funcCode_ArbRetryableTx_redeem) {
+            arbRetryable_redeem(topFrame, calldata);
+        } elseif (funcCode == const::funcCode_ArbRetryableTx_getLifetime) {
+            arbRetryable_getLifetime(topFrame, calldata);
+        } elseif (funcCode == const::funcCode_ArbRetryableTx_getTimeout) {
+            arbRetryable_getTimeout(topFrame, calldata);
+        } elseif (funcCode == const::funcCode_ArbRetryableTx_getKeepalivePrice) {
+            arbRetryable_getKeepalivePrice(topFrame, calldata);
+        } elseif (funcCode == const::funcCode_ArbRetryableTx_keepalive) {
+            arbRetryable_keepalive(topFrame, calldata);
+        } elseif (funcCode == const::funcCode_ArbRetryableTx_getBeneficiary) {
+            arbRetryable_getBeneficiary(topFrame, calldata);
+        } elseif (funcCode == const::funcCode_ArbRetryableTx_cancel) {
+            arbRetryable_cancel(topFrame, calldata);
+        } else {
+            // unrecognized function code
+            evmOp_revert_knownPc(1, 0, 0);
+        }
+    } else {
+        // this shouldn't happen -- should always be called in an EVM tx
+        evmCallStack_callHitError(27);
+    }
+}
+
+impure func arbRetryable_redeem(topFrame: EvmCallFrame, calldata: ByteArray) {  // (txId)
+    if (bytearray_size(calldata) != 36) {
+        evmOp_revert_knownPc(10, 0, 0);
+    }
+
+    // This can only be called directly from L1. Revert if called from L2.
+    if (evmCallStack_stackDepth() != 1) {
+        evmOp_revert_knownPc(11, 0, 0);
+    }
+
+    let txId = bytearray_get256(calldata, 4);
+    if (redeemRetryableTxInTx(txId)) {
```

Comment:
Seems confusing that this call doesn't include a `setRetryBufferInCurrentContext` whereas the other mutating calls do. Not a bug since it's called inside `redeemRetryableTxInTx`